keps/sig-api-machinery/0006-apply.md

```diff
@@ -213,6 +222,91 @@ Kubernetes clusters. The semantical differences between server-side
+
+  * 'f:<name>', where <name> is the name of a field in a struct, or key in a map
+  * 'v:<value>', where <value> is the exact json formatted value of a list item
+  * 'i:<index>', where <index> is position of an item in a list
+  * 'k:<keys>', where <keys> is a map of  a list item's key fields to their unique values
+
+If a path element maps to an empty set, this is a shortcut for mapping to {‘.’}
+
+This format was chosen partly because it can be stored as a RawExtension in the API, which allows it to display in json or yaml depending on what a client asks for, which makes it more human readable. But because of this there are also some major inefficiencies to this format. For example, being serialized json means that we are wasting a lot of space on control characters, for example, to represent the list item with the value ‘udp’ for it’s field ‘protocol’, we need to serialize `{“k:{\”protocol\”:\”udp\”}”:{}}`, which is a total of 29 characters, 9 are because it’s serialized JSON, 11 are just for the format of the path element, 8 are to store the field name, and only 3 are for identifying the list item uniquely.
+
+##### Areas of Improvement
+
+  1. The first round of improvements, proposed in this [PR](https://github.com/kubernetes-sigs/structured-merge-diff/pull/102), restructures the trie from nested maps to nested lists, because it permits directly serializing keys rather than encoding them as strings, which wastes much less space double-encoding strings. It also groups the path element type prefixes and whether or not self is included into a single int, by taking a path element type int (f -> 0, v -> 1, i -> 2, k -> 3) and adding either 0, 4, or 8 to specify inclusion of self, children, or both respectively. These improvements reduce the serialized size by about 25%, but there are still some obvious areas we can improve.
+
+  1. One is field names, which are now a large part of the size of the data, since we still directly serialize them as a string. To reduce the size further we will implement a statically defined string table, which maps int to strings, and instead of directly serializing common kubernetes field names, we will just write their index in this string table in base64, with a prefix of ! to specify that it refers to an item in the string table rather than an actual integer. The string table will be versioned, by prepending an int to the beginning of the top level nested list, and updated as new field names or common annotations are added. To continue supporting arbitrary labels/annotations and CRD field names, we will also allow direct field name serialization as a fallback. Without actually implementing anything, this was tested by manually replacing strings by ints in an example fieldset, and showed an improvement of an additional 30%.
```

Comment:
can you line-wrap these so it is easier to comment on things in the middle? :)

Reply:
Good thing I can like multiple times

Reply:
> prepending an int to the beginning of the top level nested list

How, exactly? It doesn't appear in the examples?

> and updated as new field names or common annotations are added

Make it clear that this will be very rare, as it will require clients (who want to understand this field) to upgrade.

Reply:
> How, exactly? It doesn't appear in the examples?

In the section for "With Improvement 1 and 2" the fieldset begins with `[1,0,!0 ...`=

Reply:
ah, ok. Easy to miss. Hm, let's see what people think about that.